src/resmom/linux/alps.c

```diff
@@ -6397,129 +6665,916 @@ alps_inventory(void)
+		log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_NODE, LOG_DEBUG, __func__,
+			log_buffer);
+		return;
+	}
+
+	if (brp_knl->data.query.type != basil_query_system) {
+		snprintf(log_buffer, sizeof(log_buffer), "Wrong query type: %d",
+			brp_knl->data.query.type);
+		log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_NODE, LOG_DEBUG, __func__,
+			log_buffer);
+		return;
+	}
+
+	if (*brp_knl->error != '\0') {
+		snprintf(log_buffer, sizeof(log_buffer), "Error in BASIL response: %s",
+			brp_knl->error);
+		log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_NODE, LOG_DEBUG, __func__,
+			log_buffer);
+		return;
+	}
+
+	sys_knl = &brp_knl->data.query.data.system;
+
+	create_vnodes_KNL(sys_knl);
+
+	free_basil_response_data(brp_knl);
+}
+
+/**
+ *
+ * @brief Create KNL vnodes.
+ * 	  'vnlp' is a global pointer that gets freed in process_hup() (mom_main.c)
+ * 	  during a MoM restart. The vnode state gets cleared and we repeat the vnode
+ *	  creation cycle i.e. creation of non-KNL vnodes (in inventory_to_vnodes())
+ *	  followed by KNL vnodes in this function.
+ *
+ * @param[in] sys_knl ALPS BASIL System Query response.
+ *
+ * @return void
+ */
+static void
+create_vnodes_KNL(basil_response_query_system_t *sys_knl)
+{
+	char	*attr, *arch;
+	char	vname[VNODE_NAME_LEN];
+	char	utilBuffer_knl[(UTIL_BUFFER_LEN * sizeof(char))];
+	int	ncpus_per_knl;
+	int	node_idx = 0;
+	int	node_count =0;
+	int	node_id = 0;
+	int	*nid_arr = NULL;
+	int	atype = READ_WRITE | ATR_DFLAG_CVTSLT;
+	char	mpphost_knl[BASIL_STRING_LONG];
+
+
+	basil_system_element_t *node_group;
+
+
+	if (sys_knl == NULL)
+		return;
+
+	strncpy(mpphost_knl, sys_knl->mpp_host, sizeof(mpphost_knl) - 1);
+	mpphost_knl[sizeof(mpphost_knl) - 1] = '\0';
+
+
+	/*
+	 * Iterate through all the Node groups in the System Query Response. Each
+	 * Node group may contain information about multiple (a 'range list') of KNL Nodes.
+	 * Each XML Element <Nodes ...> </Nodes> encapsulates information about a group of Nodes.
+	 */
+	for (node_group = sys_knl->elements; node_group; node_group = node_group->next) {
+
+		/*
+		 * The System Query XML Response contains information about KNL and
+		 * non-KNL Nodes. We are only interested in KNL nodes that are in
+		 * "batch" mode and in the "up" state.
+		 */
+		if (exclude_from_KNL_processing(node_group))
+			continue;
+
+		/*
+		 * Extract NIDs (Node ID) from node_group->nidlist.
+		 * If nidlist is empty, node_count gets set to 0 and vnode
+		 * creation in the inner for() is bypassed.
+		 */
+		nid_arr = process_nodelist_KNL(node_group->nidlist, &node_count);
+
+		/*
+		 * Create vnodes for each of the KNL Nodes listed in this Node group.
+		 * All KNL Nodes within a Node Group will have similar vnode attributes,
+		 * since all attributes in each <Nodes ...> XML element apply to all
+		 * Nodes listed in the 'rangelist'.
+		 */
+		for (node_idx = 0; node_idx < node_count; node_idx++) {
+
+			node_id = nid_arr[node_idx];
+			snprintf(vname, VNODE_NAME_LEN, "%s_%d", mpphost_knl, node_id);
```

Comment:
Once the node_id being returned is a long, then this should be updated to handle a long.